pkg/operator/minio/controller.go

```diff
@@ -0,0 +1,301 @@
+		return "", "", err
+	}
+
+	return string(val.Data["username"]), string(val.Data["password"]), nil
+}
+
+func (c *MinioController) makeMinioStatefulSet(name, namespace string, spec miniov1alpha1.ObjectStoreSpec, ownerRef meta_v1.OwnerReference) (*v1beta2.StatefulSet, error) {
+	appsClient := c.context.Clientset.AppsV1beta2()
+
+	accessKey, secretKey, err := c.getAccessCredentials(spec.Credentials.Name, spec.Credentials.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	envVars := map[string]string{
+		"MINIO_ACCESS_KEY": accessKey,
+		"MINIO_SECRET_KEY": secretKey,
+	}
+
+	podSpec := c.makeMinioPodSpec(name, namespace, minioCtrName, minioCtrImage, spec.Port, envVars, int32(spec.Storage.NodeCount), spec.ObjectStoreDataDir)
+
+	nodeCount := int32(spec.Storage.NodeCount)
+	ss := v1beta2.StatefulSet{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:            name,
+			Namespace:       namespace,
+			Labels:          map[string]string{"app": minioLabel},
+			OwnerReferences: []meta_v1.OwnerReference{ownerRef},
+		},
+		Spec: v1beta2.StatefulSetSpec{
+			Replicas: &nodeCount,
+			Selector: &meta_v1.LabelSelector{
+				MatchLabels: map[string]string{"app": minioLabel},
+			},
+			Template: podSpec,
+			VolumeClaimTemplates: []v1.PersistentVolumeClaim{
+				{
+					ObjectMeta: meta_v1.ObjectMeta{
+						Name:      minioVolumeName,
+						Namespace: namespace,
+					},
+					Spec: v1.PersistentVolumeClaimSpec{
+						AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+						Resources: v1.ResourceRequirements{
+							Requests: v1.ResourceList{
+								v1.ResourceStorage: resource.MustParse(spec.StorageSize),
+							},
+						},
+					},
+				},
+			},
+			ServiceName: name,
+			// TODO: liveness probe
+		},
+	}
+
+	return appsClient.StatefulSets(namespace).Create(&ss)
+}
+
+func (c *MinioController) makeMinioService(name, namespace string, spec miniov1alpha1.ObjectStoreSpec, ownerRef meta_v1.OwnerReference) (*v1.Service, error) {
+	coreV1Client := c.context.Clientset.CoreV1()
+
+	svc, err := coreV1Client.Services(namespace).Create(&v1.Service{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:            name,
+			Namespace:       namespace,
+			Labels:          map[string]string{"app": minioLabel},
+			OwnerReferences: []meta_v1.OwnerReference{ownerRef},
+		},
+		Spec: v1.ServiceSpec{
+			Type:     v1.ServiceTypeNodePort,
+			Selector: map[string]string{"app": minioLabel},
+			Ports:    []v1.ServicePort{{Port: spec.Port}},
+		},
+	})
+
+	return svc, err
+}
+
+func (c *MinioController) onAdd(obj interface{}) {
+	objectstore := obj.(*miniov1alpha1.ObjectStore).DeepCopy()
+
+	ownerRef := meta_v1.OwnerReference{
+		APIVersion: ObjectStoreResource.Version,
+		Kind:       ObjectStoreResource.Kind,
+		Name:       objectstore.Namespace,
+		UID:        types.UID(objectstore.ObjectMeta.UID),
+	}
+
+	// Create the headless service.
+	logger.Infof("Creating Minio headless service %s.", objectstore.Name)
```

Comment:
consider logging the namespace out too in each of these `onAdd` steps, since the operator can create a minio instance in multiple namespaces.